### PR TITLE
persist-txn: add initial metrics coverage

### PR DESCRIPTION
**Pull request by danhhz**
For now, only cover the things relevant to "eager uppers" mode.

Touches https://github.com/MaterializeInc/database-issues/issues/6685
Touches https://github.com/MaterializeInc/database-issues/issues/6888

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

I don't love the additional indentation this causes, but couldn't see an easy way around it. Definitely view the diff with whitespace suppressed

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/persist-txn/src/metrics.rs [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+impl InfallibleOpMetrics {
+    pub(crate) async fn run<R>(&self, op: impl Future<Output = R>) -> R {
+        let start = Instant::now();
+        self.started_count.inc();
+        let res = op.await;
+        self.succeeded_count.inc();
+        self.duration_seconds.inc_by(start.elapsed().as_secs_f64());
+        res
+    }
+}
```
Comment (jkosh44) [inline src/persist-txn/src/metrics.rs:156]: I _think_ `mz_ore` already has some of this functionality. Not sure if it's exactly what you want though: https://github.com/MaterializeInc/materialize/blob/068feb635e8ef16669ec000accf88b5099556a2b/src/ore/src/metrics.rs#L411-L745

**Review by jkosh44 — APPROVED** [review]
LGTM

File: src/persist-txn/src/metrics.rs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+impl Metrics {
+    /// Returns a new [Metrics] instance connected to the given registry.
+    pub fn new(registry: &MetricsRegistry) -> Self {
+        let ops = OpsMetrics::new(registry);
+        Metrics {
+            data_shard_count: registry.register(metric!(
+                name: "mz_txn_data_shard_count",
+                help: "count of data starts registered to the txn set",
```
[the remark this answers is not in the document]
Comment (danhhz) [reply src/persist-txn/src/metrics.rs]: Whoops, done!

File: src/persist-txn/src/metrics.rs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+                var_labels: ["op"],
+            )),
+            retry_count: registry.register(metric!(
+                name: "mz_txn_op_retry_count",
+                help: "count of times a txn operation retried",
+                var_labels: ["op"],
+            )),
+            succeeded_count: registry.register(metric!(
+                name: "mz_txn_op_succeeded_count",
+                help: "ount of times a txn operation succeeded",
```
[the remark this answers is not in the document]
Comment (danhhz) [reply src/persist-txn/src/metrics.rs]: Whoops, done!

File: src/persist-txn/src/metrics.rs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+                var_labels: ["op"],
+            )),
+            errored_count: registry.register(metric!(
+                name: "mz_txn_op_errored_count",
+                help: "ount of times a txn operation errored",
```
[the remark this answers is not in the document]
Comment (danhhz) [reply src/persist-txn/src/metrics.rs]: Whoops, done!

File: src/persist-txn/src/metrics.rs (hunk 1 above)
Comment (danhhz) [reply src/persist-txn/src/metrics.rs:156]: Yeah,`.wall_time` gets us the execution time, but this helper also covers counting how many times it's called, etc. Not sure reusing the ore stuff gets us a ton

**Review by danhhz — COMMENTED** [review]
TFTR!

**Pull request merged**